primitives/src/slot.rs

```diff
@@ -0,0 +1,773 @@
+            if let Some(reward_address) = &slot_address.reward_address_opt {
+                assert_eq!(reward_address, &slot_address.staker_address);
+                size += Serialize::serialize(reward_address, writer)?;
+            }
+        }
+
+        Ok(size)
+    }
+
+    fn serialized_size(&self) -> usize {
+        // Size of slot allocation
+        let mut size = SlotAllocation::SIZE;
+
+        // Size of BitVec that encodes which reward addresses are custom
+        // TODO: Replace with fixed-size BitVec
+        let reward_addresses: BitVec<BigEndian, u8> = BitVec::from_iter(self.iter()
+            .map(|item| item.reward_address_opt.is_some()));
+        size += SerializeWithLength::serialized_size::<uvar>(&reward_addresses);
+
+        for slot_address in self.iter() {
+            size += Serialize::serialized_size(&slot_address.staker_address);
+            slot_address.reward_address_opt.as_ref().map(|reward_address| {
+                size += Serialize::serialized_size(&reward_address);
+            });
+        }
+
+        size
+    }
+}
+
+impl Deserialize for StakeSlots {
+    fn deserialize<R: ReadBytesExt>(reader: &mut R) -> Result<Self, SerializingError> {
+        // Deserialize the slot allocation
+        let allocation: SlotAllocation = Deserialize::deserialize(reader)?;
+        let num_slot_addresses = allocation.num_items();
+        let allocation = allocation.as_vec();
+
+        // Deserialize which reward addresses are set
+        let reward_addresses: BitVec = DeserializeWithLength::deserialize::<uvar, R>(reader)?;
+
+        // Deserialize addresses
+        let mut slot_addresses = Vec::with_capacity(num_slot_addresses);
+
+        for (i, num_slots) in allocation.into_iter().enumerate() {
+            let staker_address: Address = Deserialize::deserialize(reader)?;
+
+            let reward_address: Option<Address> = if reward_addresses.get(i).unwrap_or_default() {
+                let address = Deserialize::deserialize(reader)?;
+                if address == staker_address {
+                    // This is an invalid encoding, as it needlessly bloats the block chain
+                    return Err(SerializingError::InvalidEncoding)
+                }
+                Some(address)
+            }
+            else {
+                None
+            };
+
+            slot_addresses.push(StakeSlotBand::new(staker_address, reward_address, num_slots));
+        }
+
+        // This invariant should always hold if SlotAllocation is implemented correctly
+        assert_eq!(slot_addresses.len(), num_slot_addresses);
+
+        Ok(Self::from_iter(slot_addresses))
+    }
+}
+
+
+/// Mapping from slot number to [SlotBand].
+///
+/// Internally this is just a [std::vec::Vec] with the indices into a [std::vec::Vec] with
+/// [SlotBand]s.
+///
+/// # ToDo
+///
+/// * Use a [nimiq_collections::segment_tree::SegmentTree]
+///
+#[derive(Clone)]
+struct SlotIndexTable {
+    index: Vec<u16>
```

Comment:
A SegmentTree might be more efficient here.

Reply:
Oh, just saw that you put the same ToDo into your comments above.

Reply:
I think for such small collections `Vec` is more efficient. It's only initialization cost and then constant lookup time.